File: cs445/threading/ruby/synchronize/mean_machine.rb
class MeanMachine
  def initialize(name, account, withdraw_amount)
    @name, @account, @withdraw_amount = name, account, withdraw_amount
  end

  def run
    deducted, remaining = nil, nil
    while deducted != 0
      deducted, remaining = @account.withdraw!(@withdraw_amount)
      print_receipt "Deducted: #{deducted}" if deducted > 0
    end
    print_receipt "Done -"
  end
private
  def print_receipt(msg) puts "#{@name} - #{msg} Remaining: #{@account.balance}"; end
end # MeanMachine

File: cs445/threading/java/src/gus/synchronize/MeanMachine.java
package gus.synchronize;

public class MeanMachine extends Thread {
  private String name;
  private Account account;
  private long amountToWithdraw;

  public MeanMachine(String name, Account account, long amountToWithdraw) {
    this.name = name;
    this.account = account;
    this.amountToWithdraw = amountToWithdraw;
    start();
  }

  public void run() {
    while(account.withdraw(amountToWithdraw)) {
      printReceipt("Deducted: " + amountToWithdraw + " Remaining");
    }
    printReceipt("Final");
  }

  private void printReceipt(String note) {
    System.out.println("machine " + name + " - " + note + ": " + account.getBalance());
  }
}
